src/time.rs

```diff
@@ -0,0 +1,135 @@
+#[derive(Resource, Clone, Copy, Debug, Hash)]
+pub struct RollbackFrameRate(pub usize);
+
+impl Default for RollbackFrameRate {
+    fn default() -> Self {
+        Self(DEFAULT_FPS)
+    }
+}
+
+/// A [`Resource`] which manages a [`Time`] directly tied to the [`GgrsFrameRate`].
+/// 
+/// This resource is available at all times, and will be updated during [`GgrsSchedule`].
+/// During this schedule, the Bevy managed [`Time`] will be replaced with one from
+/// this resource.
+#[derive(Resource, Clone)]
+pub struct RollbackTime {
+    /// An [`Instant`] which can be considered as the start time. This value will *not* be synchronized
+    /// between clients. However, the relative amount of time that has passed since this instant *will*.
+    start: Instant,
+    /// Our internally managed [`Time`], which will update exactly once each frame, passing exactly `1/fps` seconds.
+    time: Time,
+    /// Temporary storage location for an existing [`Time`] managed by Bevy.
+    bevy_time: Option<Time>,
+}
+
+impl Deref for RollbackTime {
+    type Target = Time;
+
+    fn deref(&self) -> &Self::Target {
+        &self.time
+    }
+}
+
+impl Default for RollbackTime {
+    fn default() -> Self {
+        let now = Instant::now();
+
+        Self {
+            start: now,
+            time: Time::new(now),
+            bevy_time: None,
+        }
+    }
+}
+
+/// Provides a rollback-aware [`Time`] resource, suitable for use within the [`GgrsSchedule`].
+/// To ensure it is available and up-to-date, you should add time-sensitive systems
+/// to the [`AdvanceWorldSet::Main`] set.
+pub struct RollbackTimePlugin;
+
+impl RollbackTimePlugin {
+    /// Update the [`RollbackTime`] managed by this plugin.
+    pub fn update(
+        mut ggrs_time: ResMut<RollbackTime>,
+        frame: Res<RollbackFrameCount>,
+        framerate: Res<RollbackFrameRate>,
+    ) {
+        let this_frame = TryInto::<u64>::try_into(frame.0).unwrap_or_default();
+
+        // 1_000_000_000 fits within a u32, and so does frame, making their product at most u64 in size
+        let runtime_nanos = this_frame * 1_000_000_000 / (framerate.0 as u64);
+
+        // By scaling to nanoseconds, rounding error should be minimised.
+        let runtime = Duration::from_nanos(runtime_nanos);
+
+        // Absolute formula avoids accumulating error.
+        let now = ggrs_time.start + runtime;
+
+        ggrs_time.time.update_with_instant(now);
+    }
+
+    /// Insert a [`Time`] from the [`RollbackTime`] managed by this plugin.
+    pub fn insert(
+        mut commands: Commands,
+        mut ggrs_time: ResMut<RollbackTime>,
+        bevy_time: Option<ResMut<Time>>,
+    ) {
+        if let Some(mut bevy_time) = bevy_time {
+            let mut time = ggrs_time.time.clone();
+
+            std::mem::swap(bevy_time.as_mut(), &mut time);
+
+            ggrs_time.bevy_time = Some(time);
+        } else {
+            commands.insert_resource(ggrs_time.time.clone());
```

Comment:
Maybe modify the resource directly to avoid 1-frame-delay problems, and use `time.clone_from(ggrs_time.time)`?

Reply:
I do agree, we can always assume there is a `Time` resource to overwrite, since GGRS already relies on its existence! As for `clone_from`, I use `swap` here to ensure I never actually overwrite the `Time` value managed by Bevy, only replace it during the `GgrsSchedule`